src/connector/src/sink/file_sink/opendal_sink.rs

```diff
@@ -0,0 +1,315 @@
+    type Coordinator = DummySinkCommitCoordinator;
+    type LogSinker = LogSinkerOf<OpenDalSinkWriter>;
+
+    const SINK_NAME: &'static str = S::SINK_NAME;
+
+    async fn validate(&self) -> Result<()> {
+        if !self.is_append_only {
+            return Err(SinkError::Config(
+                anyhow!("File only supports append-only mode at present, please change the query to append-only, or use `force_append_only = 'true'`")
+            ));
+        }
+        Ok(())
+    }
+
+    async fn new_log_sinker(
+        &self,
+        writer_param: crate::sink::SinkWriterParam,
+    ) -> Result<Self::LogSinker> {
+        Ok(OpenDalSinkWriter::new(
+            self.op.clone(),
+            &self.path,
+            self.schema.clone(),
+            self.is_append_only,
+            writer_param.executor_id,
+            self.format_desc.encode.clone(),
+        )?
+        .into_log_sinker(writer_param.sink_metrics))
+    }
+}
+
+impl<S: OpendalSinkBackend> TryFrom<SinkParam> for FileSink<S> {
+    type Error = SinkError;
+
+    fn try_from(param: SinkParam) -> std::result::Result<Self, Self::Error> {
+        let schema = param.schema();
+        let config = S::from_btreemap(param.properties)?;
+        let path = S::get_path(&config);
+        let op = S::new_operator(config)?;
+
+        Ok(Self {
+            op,
+            path,
+            schema,
+            is_append_only: param.sink_type.is_append_only(),
+            format_desc: param
+                .format_desc
+                .ok_or_else(|| SinkError::Config(anyhow!("missing FORMAT ... ENCODE ...")))?,
+            marker: PhantomData,
+        })
+    }
+}
+
+pub struct OpenDalSinkWriter {
+    schema: SchemaRef,
+    operator: Operator,
+    sink_writer: Option<FileWriterEnum>,
+    is_append_only: bool,
+    write_path: String,
+    epoch: Option<u64>,
+    executor_id: u64,
+    encode_type: SinkEncode,
+}
+
+/// The `FileWriterEnum` enum represents different types of file writers used for various sink
+/// implementations.
+///
+/// # Variants
+///
+/// - `ParquetFileWriter`: Represents a Parquet file writer using the `AsyncArrowWriter<W>`
+/// for writing data to a Parquet file. It accepts an implementation of W: `AsyncWrite` + `Unpin` + `Send`
+/// as the underlying writer. In this case, the `OpendalWriter` serves as the underlying writer.
+///
+/// - `FileWriter`: Represents a file writer for sinks other than Parquet. It uses the `OpendalWriter`
+/// directly for writing data to the file.
+///
+/// The choice of writer used during the actual writing process depends on the encode type of the sink.
+enum FileWriterEnum {
+    ParquetFileWriter(AsyncArrowWriter<Compat<FuturesAsyncWriter>>),
+    // `FileWriter` is used to write files other than parquet format to the file system.
+    // Since only sinks in parquet format are currently supported, its implementation is temporarily commented out.
+
+    // FileWriter(OpendalWriter)
```

Comment:
Given that we don't have any struct called `FileWrtier` in the code bases, I suggest we remove this line instead of commenting it out.

Reply:
`FileWriter` is for further json/csv encode, I leave this is to explain the future form of `FileWriterEnum`. Since they are not introduced in this pr, let me remove it first.

Reply:
But I still prefer to keep `FileWriterEnum` as an enum, not a struct to facilitate adding other encode in the future.